### PR TITLE
GPU host requirement proposal (devcontainers/spec#82)

### PR DESCRIPTION
**Pull request by chrmarti**

### CODE REVIEW

**Review by Chuxel — APPROVED** [review]

**Comment by l0rem1psum** [comment]
Based on my current workflow and use cases, I feel that it will also be useful to include GPU architecture as one of the requirement fields. In the case of NVIDIA GPUs, for instance, we can indicate whether T4 or P4 GPUs. It will also be useful to indicate the number of GPUs required.

**Comment by chrmarti** [comment]
@l0rem1psum Number of different GPUs? (We support number of cores.) It might make sense to first collect feedback from the use of the currently proposed approach and then extend it to multiple GPUs and specific GPU types. (Although "GPU types" sounds like a constantly changing set of identifiers and capabilities which we might not want to track in a specification.)

**Comment by l0rem1psum** [comment]
There might be applications where multiple GPUs of different architectures/models are required, but I have only worked with multiple GPUs of the same configuration. On a side note, using the term "cores" on GPUs may oftentimes cause confusion as far as I understand since a single GPU has hundreds to thousands of cores.

**Comment by l0rem1psum** [comment]
Not sure about other users, the few requirements I often care about are as follow:

1. Number of GPU units
2. GPU architecture
3. GPU memory
4. GPU compute capability 
5. GPU driver version
6. CUDA version (not really a host requirement but is bound by driver version)

**Review by joshspicer — APPROVED** [review]
This sounds reasonable to me.  The additional comments on this PR indicate additional properties that would be useful that I don't have GPU experience to speak to those comments

File: proposals/gpu-host-requirement.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# GPU host requirement
+
+## Goal
+
+This proposal adds support for GPU to the existing support for host requirements.
```
Comment (bamurtaugh) [inline proposals/gpu-host-requirement.md:5]: Small wording suggestion:

```suggestion
This proposal adds GPU support to the existing host requirements properties.
```

File: proposals/gpu-host-requirement.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+- Community PR: https://github.com/devcontainers/cli/pull/173
+
+## Proposal
+
+We propose to add a new `gpu` property to the `hostRequirements` object in the devcontainer.json schema. The property can be a boolean value, the string `optional` or an object:
+
+(Additional row for the existing table in the spec.)
+| Property | Type | Description |
+|----------|------|-------------|
+| `hostRequirements.gpu` 🏷️ | boolean \| 'optional' | Indicates whether or not a GPU is required. A value 'optional' indicates that might be used if available. The object is described in the table below. The default is false. For example: `"hostRequirements": "optional"` |
```
Comment (bamurtaugh) [inline proposals/gpu-host-requirement.md:16]:
```suggestion
| `hostRequirements.gpu` 🏷️ | boolean | 'optional' | Indicates whether or not a GPU is required. A value 'optional' indicates that might be used if available. The object is described in the table below. The default is false. For example: `"hostRequirements": "optional"` |
```

Comment (chrmarti) [reply proposals/gpu-host-requirement.md:16]: The backslash is needed to render the pipe symbol. Without the backslash the pipe separates two columns. Thanks.

**Comment by chrmarti** [comment]
Thanks @l0rem1psum, I have opened a new issue to track your suggestions.

**Pull request merged**